#1312: metadata for read-only arrays

mo-lottieturner opened the pull request:
.

mo-lottieturner commented:
#1312

mo-lottieturner commented:
tests all green except codecov

mo-lottieturner commented:
Just documentation left to do

mo-lottieturner commented:
just did a double check of pycodestyle and pylint, after noticing the above modifier line had slipped through - pycodestyle clear in `src/psyclone`, but pylint has quite a few 'Access to protected member' in the test files - should I ignore these?
```
tests/domain/lfric/kernel/array_arg_metadata_test.py:78:36: W0212: Access to a protected member _get_metadata of a client class (protected-access)

tests/domain/lfric/kernel/common_meta_arg_metadata_test.py:126:11: W0212: Access to a protected member _datatype of a client class (protected-access)
tests/domain/lfric/kernel/common_meta_arg_metadata_test.py:127:11: W0212: Access to a protected member _access of a client class (protected-access)
tests/domain/lfric/kernel/common_meta_arg_metadata_test.py:158:11: W0212: Access to a protected member _datatype of a client class (protected-access)
tests/domain/lfric/kernel/common_meta_arg_metadata_test.py:159:11: W0212: Access to a protected member _access of a client class (protected-access)
tests/domain/lfric/kernel/common_meta_arg_metadata_test.py:299:11: W0212: Access to a protected member _datatype of a client class (protected-access)
tests/domain/lfric/kernel/common_meta_arg_metadata_test.py:303:11: W0212: Access to a protected member _access of a client class (protected-access)

tests/domain/lfric/lfric_array_mdata_test.py:87:16: W0212: Access to a protected member _inits of a client class (protected-access)
tests/domain/lfric/lfric_array_mdata_test.py:89:8: W0212: Access to a protected member _init_array of a client class (protected-access)
tests/domain/lfric/lfric_array_mdata_test.py:160:17: W0212: Access to a protected member _inits of a client class (protected-access)
tests/domain/lfric/lfric_array_mdata_test.py:169:8: W0212: Access to a protected member _init_scalar of a client class (protected-access)
```

arporter commented:
> but pylint has quite a few 'Access to protected member' in the test files - should I ignore these?

For the tests, `pylint` should be run in the `tests` directory. That way it picks up some test-specific settings (one of which is to turn off such warnings).

arporter commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -1343,6 +1378,15 @@ checks (when generating the PSy layer) that any kernels which read
 
+.. _lfric-array-sizes:
+
+Array sizes
```
I think this should be included in the same example where the fourth and fifth entries for fields and operators are described.

Also, I may be missing something here but why `NRANKS*n`? I think that is confusing as it mixes the notation used for a field vector with something different. Can we not just have "n"? I suppose we could have `GH_ARRAY*n` by analogy with a field vector and do away with this extra entry altogether. In fact, maybe it should be `GH_SCALAR*n` as that really does then line up with the way we do field vectors?

mo-lottieturner replied:
For posterity: the reason behind not using `GH_ARRAY*n` (or `GH_SCALAR*n`) is because `n` here does not represent the number of rows in a vector, but the number of dimensions in an array (so `n=1` is a vector, `n=2` is a 2d-array, etc)

arporter commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -112,13 +116,13 @@ Please see the :ref:`algorithm-layer` section for a description of the
-which provides information required by a kernel to operate on fields
-(see section :ref:`dynamo0.3-quadrature` for more details).
+six LFRic (Dynamo 0.3) API argument types: :ref:`scalar <lfric-scalar>`,
+:ref:`array <lfric-array>`,:ref:`field <lfric-field>`, :ref:`field vector
```
This isn't rendering correctly for me:
![image](https://github.com/stfc/PSyclone/assets/17409212/56d4f759-fc2d-4ece-b92c-f9777d25bcae)

mo-lottieturner replied:
missing space issue, should be fixed

arporter commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -423,6 +437,20 @@ associated kernel metadata description and their precision:
 +--------------------------+---------------------------------+-----------+
```
Please mention arrays at L409.

arporter commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -423,6 +437,20 @@ associated kernel metadata description and their precision:
 +--------------------------+---------------------------------+-----------+
 | LOGICAL(L_DEF)           | GH_SCALAR, GH_LOGICAL           | L_DEF     |
 +--------------------------+---------------------------------+-----------+
+| REAL(R_DEF)              | GH_ARRAY, GH_REAL               | R_DEF     |
```
I think these new lines can be merged with the corresponding lines for scalars by altering the latter to say "GH_SCALAR/GH_ARRAY, GH_INTEGER" etc. In fact, once we've discussed my suggestion of `GH_SCALAR*n` these changes might not be required at all.

arporter reviewed (changes requested):
Thanks for this @mo-lottieturner. I should have said, but I've come to realise that for something like this it's always best to just do the documentation first and then, once we're all agreed, proceed to the implementation.

As it is, I've only looked at the documentation so far. I'm not too sure about the form of the new metadata as it seems to me to be a bit more complex than is necessary (although I'm happy to be corrected). I think there are two choices, extend `GH_SCALAR` so that we also support `GH_SCALAR*n`. This has the advantage that it follows what we do for field vectors and is therefore intuitive. It also reduces the confusion about why a field is not an array. Failing that, my second suggestion would be to simplify the fourth metadata entry for a `GH_ARRAY` so that it is just `nranks` - there's no need to use the `*` notation.

Happy to discuss with all interested parties :-)

mo-lottieturner commented:
For posterity: If memory serves, we agreed to stick with `GH_ARRAY`, but that `NRANKS*n` can be switched to just `n`

arporter commented on src/psyclone/domain/lfric/kernel/array_arg_metadata.py:
```diff
@@ -0,0 +1,159 @@
+
+'''Module containing the ArrayArgMetadata class which captures the metadata
+associated with an array argument. Supports the creation, modification
+and Fortran output of an Array argument.
```
Sorry to be a pain but I keep thinking we need to qualify our mentions of  'array' to make it clear it is something special (after all, a field boils down to an array and I think it could lead to confusion). Therefore, I'm wondering whether `ScalarArray` (and so `GH_SCALAR_ARRAY`) would be a better name for this concept - it signals that we really do mean something other than a field?
Something to discuss with at least @TeranIvy and possibly others.

TeranIvy replied:
I think you have a point there, @arporter, as we can have arrays of fields.

TeranIvy replied:
Discussed this with @mo-lottieturner and she's agreed to change it.

mo-lottieturner replied:
does this include changing the file name from array_arg_metadata to scalar_array_arg_metadata?

mo-lottieturner replied:
@arporter could I have an answer to this please?

arporter replied:
If the class name has changed then yes please.

mo-alistairp replied:
I believe I've changed all occurances of array to ScalarArray and I've changed file names

arporter commented on src/psyclone/domain/lfric/kernel/array_arg_metadata.py:
```diff
@@ -0,0 +1,159 @@
+    form_arg_index = 0
+    datatype_arg_index = 1
+    access_arg_index = 2
+    array_size_arg_index = 3
```
I suggest replacing "size" with "rank" here.

mo-lottieturner replied:
this is an artefact of distinguishing between `NRANKS*<n>` and `n` - I'll have a look at what else might need changing of this sort

mo-lottieturner replied:
I've changed this to `array_ndims_arg_index` to match with everything else that refers to this argument, including the property

arporter commented on src/psyclone/domain/lfric/kernel/array_arg_metadata.py:
```diff
@@ -0,0 +1,159 @@
+    check_name = "array"
+    # The number of arguments in the language-level metadata (min and
+    # max values).
+    nargs = (4)
```
Should be `(4, 4)` perhaps but I guess I'll find out later.

arporter commented on src/psyclone/domain/lfric/kernel/array_arg_metadata.py:
```diff
@@ -0,0 +1,159 @@
+
+    :param str datatype: the datatype of this array (GH_INTEGER, ...).
+    :param str access: the way the kernel accesses this array (GH_READ).
+    :param str function_space: the function space that this array is \
```
This should be documenting `array_ndims`.

arporter commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -1388,6 +1410,15 @@ checks (when generating the PSy layer) that any kernels which read
+where *n > 0* is the number of Fortran ranks representing the dimension of the
+array.
+
 .. _lfric-function-space:
```
The rules for the arguments to a Kernel subroutine also need to be updated. Presumably, an Array argument will need to be followed by `ndims` and then possibly an integer, rank-1 array of size `ndims` containing the size of each dimension? Or, will `ndims` always be a literal (known) value such that we can each of the dimension sizes as an explicit subroutine argument? (This will allow us to explicitly declare the shape of the `array` argument in the PSy layer which is something we always aim to do.)

TeranIvy replied:
Yes, we need to pass `ndims` and rank-1 array containing sizes. The proposed design is in the description of #1312, together with examples of code for Kernel, Algorithm and PSy layers (all of them are compilable as far as I remember).

arporter replied:
@mo-alistairp this one is still outstanding - it's https://psyclone.readthedocs.io/en/latest/user_guide/lfric.html#rules-for-general-purpose-kernels that needs to be updated.

arporter commented on src/psyclone/domain/lfric/kernel/array_arg_metadata.py:
```diff
@@ -0,0 +1,159 @@
+        '''
+        :param str value: set the function space to the
+            specified value.
+        '''
```
Please document (after a blank line) the various :raises:

arporter commented on src/psyclone/domain/lfric/kernel/array_arg_metadata.py:
```diff
@@ -0,0 +1,159 @@
+        return (f"arg_type({self.form}, {self.datatype}, {self.access}, "
+                f"{self.array_ndims})")
+
+    @staticmethod
```
Unless I'm missing something, this whole method can be removed as it's identical to the one in the class from which this one inherits.

arporter reviewed (changes requested):
Thanks for those changes @mo-lottieturner. The docs are looking pretty good now although I realise that the section describing the detailed rules for the subroutine arguments of a kernel also need to be updated.
I've now gone through the code changes but not the test files. Generally speaking it looks good although you'll see that I'm suggesting (another?) name change for clarity as I keep getting confused about distinguishing an 'array' from a 'field'. We'll need to agree any name change with at least @TeranIvy and whoever else she thinks might care!

mo-lottieturner commented on src/psyclone/domain/lfric/lfric_arg_descriptor.py:
```diff
@@ -270,43 +275,83 @@ def _validate_vector_size(self, separator, arg_type):
+                f"but '{arg_type.args[3]}' was found in "
+                f"'{arg_type}'.") from err
+
+        # ... or it is less than 1 (1 is the default for all fields)...
```
This section was copypasted from _validate_vector_size - the same comment is there (less than 2, 1 the default). Not sure I follow it either - should I remove the brackets section?

arporter replied:
I think we just remove the brackets section.

mo-alistairp replied:
Removed the bracketed section

mo-lottieturner commented on src/psyclone/domain/lfric/lfric_arg_descriptor.py:
```diff
@@ -599,7 +644,65 @@ def _init_scalar(self, arg_type):
+
+    def _init_array(self, arg_type):
+        '''
+        Validates metadata descriptors for scalar array arguments and
```
as in ScalarArray instead of scalar array?

mo-lottieturner replied:
It feels inconsistent for scalar arrays to be the only ones to use camel case

mo-lottieturner replied:
@arporter and an answer to this too please

arporter replied:
In this case, I think "ScalarArray" is what we want as this indicates it is a special, named quantity in the API.

mo-alistairp replied:
I believe I've changed all the occurances of array to ScalarArray

mo-lottieturner commented on src/psyclone/domain/lfric/lfric_arg_descriptor.py:
```diff
@@ -599,7 +644,65 @@ def _init_scalar(self, arg_type):
+                               passed in.
+        :raises ParseError: if there are not exactly 4 metadata arguments.
+        :raises InternalError: if an array argument has an invalid data type.
+        :raises ParseError: if array arguments do not have read-only access.
```
not sure this is right - think it should be 'if the array argument has an invalid access type'

arporter replied:
Having made my way through this review I've now remembered that read-only *is* the only valid access type so this should read "if the array argument does not have read-only access".

mo-lottieturner commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -1242,6 +1273,8 @@ the :ref:`LFRic fields <lfric-field>`):
 +========================+=================================+
```
I'm not clear on what needs changing where for this comment, sorry

mo-lottieturner replied:
I think I've corrected what needs correcting here

mo-lottieturner commented on src/psyclone/domain/lfric/kernel/common_meta_arg_metadata.py:
```diff
@@ -200,6 +202,28 @@ def get_vector_length(cls, fparser2_tree):
+    @classmethod
+    def get_array_ndims(cls, fparser2_tree):
+        '''Retrieves the array ndims metadata value found within the
+        supplied fparser2 tree and checks that it is valid.
```
The validation is done in the setter - should some or all of these be moved to the getter?

mo-lottieturner commented:
@arporter not all review comments have been addressed yet, but as I'm handing this over to Alistair, it would be helpful for you to mark the addressed comments as resolved, if you have a moment?

arporter commented on src/psyclone/domain/lfric/kernel/array_arg_metadata.py:
```diff
@@ -101,59 +100,53 @@ def fortran_string(self):
+        :returns: the array ndims value extracted from the fparser2 tree.
+        :rtype: str
+
+        :raises TypeError: if the array ndims is not a string.
```
This method only actually raises ValueErrors. Also, I think this checking should probably be done in the array_ndims setter. I'm not sure whether we should duplicate - we'd need to look at what we do for other properties.

mo-alistairp replied:
Hi @arporter, I've had a little look into this and I think I've understood (apologies if not!). 

I agree that it makes more sense to validate this in the setter rather than the getter. I had a little look, and it seems in both `field_vector_arg_metadata.py` and `evaluator_targets_metadata.py` the setter validates the data (although admittedly neither seem to have a specific getter function from the fparser2 tree). 

When you talk about duplication, do you mean validating the data in both the getter and setter?

mo-alistairp replied:
As an extension to what I said above, I;ve just noticed that in `inter_grid_arg_metadata.py` there is some checking in both `@mesh_arg.setter` and in the static `get_mesh_arg` (which gets the value from the fparser2 tree), which seems more similar to our case here. So I'm not enirely sure what the best course of action is.

arporter replied:
In that case, I think we live with the duplication of the validation code and do it in both. Perhaps just a little utility method unless that's over the top.

arporter commented:
OK @mo-lottieturner and @mo-alistairp, I've gone through and resolved everything that has been done.

mo-alistairp commented on src/psyclone/domain/lfric/lfric_arg_descriptor.py:
```diff
@@ -599,7 +644,65 @@ def _init_scalar(self, arg_type):
+        if self._data_type not in const.VALID_ARRAY_DATA_TYPES:
+            raise InternalError(
+                f"Expected one of {const.VALID_ARRAY_DATA_TYPES} as the "
+                f"array data type but got '{self._data_type}'.")
```
Can I check, with this and similar comments (s/size/rank/) are you wanting me to simply replace those exactly?

mo-alistairp commented on src/psyclone/domain/lfric/lfric_arg_descriptor.py:
```diff
@@ -713,6 +820,19 @@ def vector_size(self):
+    @property
+    def array_ndims(self):
+        '''
+        Returns the array size of the argument. This will be 1 if ``*n``
```
Done

mo-alistairp commented:
I've addrsesed all the open comments. It's possible I've been a little overzealous with the changes from array to ScalarArray, so just let me know. Otherwise, this is ready for review

arporter commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -1388,6 +1410,15 @@ checks (when generating the PSy layer) that any kernels which read
+Array sizes
+^^^^^^^^^^^
+
+The size of a :ref:`scalar array <lfric-array>` is described by ``NRANKS*<n>``,
```
This just needs to be updated with an example now.

arporter reviewed (changes requested):
Thanks for taking this on @mo-alistairp. It's getting there now.

Somewhere in my comments I requested that you add an "e.g." before GH_READ but I now realise that GH_READ is the only permitted access for a ScalarArray so please don't  make that change. In fact, possibly we should emphasise that restriction instead.

I've realised that we don't have an example of the use of this new kernel argument but I suggest you do that as a separate issue/pr so that we can get this one over the line. Perhaps add it as something to do in #1312 and make this a 'towards' PR.

I'll fire off the integration tests just to be on the safe side.

mo-alistairp commented on src/psyclone/domain/lfric/kernel/common_meta_arg_metadata.py:
```diff
@@ -186,7 +187,7 @@ def get_vector_length(cls, fparser2_tree):
         :returns: the vector length value extracted from the fparser2 tree.
         :rtype: str
```
Reverted the changes

arporter commented on doc/user_guide/lfric.rst:
```diff
@@ -2062,6 +2062,9 @@ conventions, are:
       from the metadata (see :ref:`lfric-api-meta-args`).
+   5) If the current entry is a ScalarArray then include the Fortran
+      variable in the argument list. The ScalarArray must be denoted
+      with intent ``in`` to match its read-only nature.
```
Apologies, I found a previous comment where @TeranIvy spelt-out what needed to be done. The specification is here https://github.com/stfc/PSyclone/pull/2173/files - for each ScalarArray argument we need to pass an additional integer array holding the size of each dimension of the array.

mo-alistairp replied:
I was using that doumentation, but just misread - so apologies from me too! That should be updated now.

mo-alistairp commented on doc/user_guide/lfric.rst:
```diff
@@ -1399,6 +1425,15 @@ checks (when generating the PSy layer) that any kernels which read
+
+The size of a :ref:`scalar array <lfric-array>` is described by ``<n>``,
+where *n > 0* is the number of Fortran ranks representing the dimension of the
+array.
```
I can't seem to respond to each individual comment, but I believe I've addressed each of your requests. I've -

- changed it over so it saves `array_ndims` as an `int`, and changed the relavent tests and documentation to reflect that 
- updated a couple of other functions to have typehinting 
- added the example you requested
- fixed a couple of typos
